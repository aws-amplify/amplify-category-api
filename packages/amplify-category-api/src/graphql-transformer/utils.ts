import fs from 'fs-extra';
import * as path from 'path';
import {
  DeploymentResources,
  AmplifyApiGraphQlResourceStackTemplate,
  TransformerResolverProvider,
  AppsyncStackCommon,
  FunctionDirectiveStack,
  HttpsDirectiveStack,
  ModelDirectiveStack,
  OpenSearchDirectiveStack,
} from '@aws-amplify/graphql-transformer-interfaces';
import { TransformerProjectConfig, SyncUtils } from '@aws-amplify/graphql-transformer-core';
import rimraf from 'rimraf';
import {
  $TSContext, AmplifyCategories, CloudformationProviderFacade, JSONUtilities, pathManager, stateManager,
} from '@aws-amplify/amplify-cli-core';
import { CloudFormation, Fn } from 'cloudform-types';
import { ResourceConstants } from 'graphql-transformer-common';
import { pullAllBy, find } from 'lodash';
import { printer } from '@aws-amplify/amplify-prompts';
import { CfnResource } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import _ from 'lodash';
import { ConstructResourceMeta } from './types';


const PARAMETERS_FILE_NAME = 'parameters.json';
const CUSTOM_ROLES_FILE_NAME = 'custom-roles.json';
const AMPLIFY_ADMIN_ROLE = '_Full-access/CognitoIdentityCredentials';
const AMPLIFY_MANAGE_ROLE = '_Manage-only/CognitoIdentityCredentials';
const PROVIDER_NAME = 'awscloudformation';

interface CustomRolesConfig {
  adminRoleNames?: Array<string>;
}

export const getIdentityPoolId = async (ctx: $TSContext): Promise<string | undefined> => {
  const { allResources, resourcesToBeDeleted } = await ctx.amplify.getResourceStatus('auth');
  const authResources = pullAllBy(allResources, resourcesToBeDeleted, 'resourceName');
  const authResource = find(authResources, { service: 'Cognito', providerPlugin: PROVIDER_NAME }) as any;
  return authResource?.output?.IdentityPoolId;
};

export const getAdminRoles = async (ctx: $TSContext, apiResourceName: string | undefined): Promise<Array<string>> => {
  let currentEnv;
  const adminRoles = new Array<string>();

  try {
    currentEnv = ctx.amplify.getEnvInfo().envName;
  } catch (err) {
    // When there is no environment info, return [] - This is required for sandbox pull
    return [];
  }

  // admin ui roles
  try {
    const amplifyMeta = stateManager.getMeta();
    const appId = amplifyMeta?.providers?.[PROVIDER_NAME]?.AmplifyAppId;
    const res = await CloudformationProviderFacade.isAmplifyAdminApp(ctx, appId);
    if (res.userPoolID) {
      adminRoles.push(`${res.userPoolID}${AMPLIFY_ADMIN_ROLE}`, `${res.userPoolID}${AMPLIFY_MANAGE_ROLE}`);
    }
  } catch (err) {
    // no need to error if not admin ui app
  }

  // additonal admin role checks
  if (apiResourceName) {
    // lambda functions which have access to the api
    const { allResources, resourcesToBeDeleted } = await ctx.amplify.getResourceStatus('function');
    const resources = pullAllBy(allResources, resourcesToBeDeleted, 'resourceName')
      .filter((r: any) => r.dependsOn?.some((d: any) => d?.resourceName === apiResourceName))
      .map((r: any) => `${r.resourceName}-${currentEnv}`);
    adminRoles.push(...resources);

    // check for custom iam admin roles
    const customRoleFile = path.join(
      pathManager.getResourceDirectoryPath(undefined, AmplifyCategories.API, apiResourceName),
      CUSTOM_ROLES_FILE_NAME,
    );
    if (fs.existsSync(customRoleFile)) {
      const customRoleConfig = JSONUtilities.readJson<CustomRolesConfig>(customRoleFile);
      if (customRoleConfig && customRoleConfig.adminRoleNames) {
        const adminRoleNames = customRoleConfig.adminRoleNames
          // eslint-disable-next-line no-template-curly-in-string
          .map((r) => (r.includes('${env}') ? r.replace('${env}', currentEnv) : r));
        adminRoles.push(...adminRoleNames);
      }
    }
  }
  return adminRoles;
};

export function mergeUserConfigWithTransformOutput(
  userConfig: TransformerProjectConfig,
  transformOutput: DeploymentResources,
  opts?: any,
): DeploymentResources {
  const userFunctions = userConfig.functions || {};
  const userResolvers = userConfig.resolvers || {};
  const userPipelineFunctions = userConfig.pipelineFunctions || {};
  const { functions } = transformOutput;
  const { resolvers } = transformOutput;
  const { pipelineFunctions } = transformOutput;

  if (!opts?.disableFunctionOverrides) {
    for (const userFunction of Object.keys(userFunctions)) {
      functions[userFunction] = userFunctions[userFunction];
    }
  }

  if (!opts?.disablePipelineFunctionOverrides) {
    const pipelineFunctionKeys = Object.keys(userPipelineFunctions);

    if (pipelineFunctionKeys.length > 0) {
      printer.warn(
        ' You are using the "pipelineFunctions" directory for overridden and custom resolvers. '
          + 'Please use the "resolvers" directory as "pipelineFunctions" will be deprecated.\n',
      );
    }

    for (const userPipelineFunction of pipelineFunctionKeys) resolvers[userPipelineFunction] = userPipelineFunctions[userPipelineFunction];
  }

  if (!opts?.disableResolverOverrides) {
    for (const userResolver of Object.keys(userResolvers)) {
      if (userResolver !== 'README.md') {
        resolvers[userResolver] = userResolvers[userResolver].toString();
      }
    }
  }

  const stacks = overrideUserDefinedStacks(userConfig, transformOutput);

  return {
    ...transformOutput,
    functions,
    resolvers,
    pipelineFunctions,
    stacks,
  };
}

function overrideUserDefinedStacks(userConfig: TransformerProjectConfig, transformOutput: DeploymentResources) {
  const userStacks = userConfig.stacks || {};
  const { stacks, rootStack } = transformOutput;

  const resourceTypesToDependOn = {
    'AWS::CloudFormation::Stack': true,
    'AWS::AppSync::GraphQLApi': true,
    'AWS::AppSync::GraphQLSchema': true,
  };

  const allResourceIds = Object.keys(rootStack.Resources).filter((k: string) => {
    const resource = rootStack.Resources[k];
    return resourceTypesToDependOn[resource.Type];
  });

  const parametersKeys = Object.keys(rootStack.Parameters);
  const customStackParams = parametersKeys.reduce(
    (acc: any, k: string) => ({
      ...acc,
      [k]: Fn.Ref(k),
    }),
    {},
  );

  customStackParams[ResourceConstants.PARAMETERS.AppSyncApiId] = Fn.GetAtt(ResourceConstants.RESOURCES.GraphQLAPILogicalID, 'ApiId');

  const updatedParameters = rootStack.Parameters;

  for (const userStack of Object.keys(userStacks)) {
    if (stacks[userStack]) {
      throw new Error(`You cannot provide a stack named ${userStack} as it \
            will be overwritten by a stack generated by the GraphQL Transform.`);
    }
    const userDefinedStack = userStacks[userStack];

    for (const key of Object.keys(userDefinedStack.Parameters)) {
      if (customStackParams[key] == null) {
        customStackParams[key] = Fn.Ref(key);

        if (updatedParameters[key]) throw new Error(`Cannot redefine CloudFormation parameter ${key} in stack ${userStack}.`);
        else updatedParameters[key] = userDefinedStack.Parameters[key];
      }
    }

    const parametersForStack = Object.keys(userDefinedStack.Parameters).reduce(
      (acc, k) => ({
        ...acc,
        [k]: customStackParams[k],
      }),
      {},
    );

    stacks[userStack] = userDefinedStack;

    const stackResourceId = userStack.split(/[^A-Za-z]/).join('');
    const customNestedStack = new CloudFormation.Stack({
      Parameters: parametersForStack,
      TemplateURL: Fn.Join('/', [
        'https://s3.amazonaws.com',
        Fn.Ref(ResourceConstants.PARAMETERS.S3DeploymentBucket),
        Fn.Ref(ResourceConstants.PARAMETERS.S3DeploymentRootKey),
        'stacks',
        userStack,
      ]),
    }).dependsOn(allResourceIds);
    rootStack.Resources[stackResourceId] = customNestedStack;
  }

  rootStack.Parameters = updatedParameters;

  return stacks;
}

/**
 * Writes a deployment to disk at a path.
 */
export async function writeDeploymentToDisk(
  context: $TSContext,
  deployment: DeploymentResources,
  directory: string,
  rootStackFileName = 'rootStack.json',
  buildParameters: Object,
) {
  fs.ensureDirSync(directory);
  // Delete the last deployments resources except for tsconfig if present
  emptyBuildDirPreserveTsconfig(directory);

  // Write the schema to disk
  const { schema } = deployment;
  const fullSchemaPath = path.normalize(`${directory}/schema.graphql`);
  fs.writeFileSync(fullSchemaPath, schema);

  // Setup the directories if they do not exist.
  initStacksAndResolversDirectories(directory);

  // Write resolvers to disk
  const resolverFileNames = Object.keys(deployment.resolvers);
  const resolverRootPath = resolverDirectoryPath(directory);
  for (const resolverFileName of resolverFileNames) {
    const fullResolverPath = path.normalize(`${resolverRootPath}/${resolverFileName}`);
    fs.writeFileSync(fullResolverPath, deployment.resolvers[resolverFileName]);
  }

  // Write pipeline resolvers to disk
  const pipelineFunctions = Object.keys(deployment.pipelineFunctions);
  const pipelineFunctionRootPath = pipelineFunctionDirectoryPath(directory);
  for (const functionFileName of pipelineFunctions) {
    const fullTemplatePath = path.normalize(`${pipelineFunctionRootPath}/${functionFileName}`);
    fs.writeFileSync(fullTemplatePath, deployment.pipelineFunctions[functionFileName]);
  }

  // Write the stacks to disk
  const stackNames = Object.keys(deployment.stacks);
  const stackRootPath = stacksDirectoryPath(directory);
  for (const stackFileName of stackNames) {
    const fileNameParts = stackFileName.split('.');
    if (fileNameParts.length === 1) {
      fileNameParts.push('json');
    }
    const fullFileName = fileNameParts.join('.');
    throwIfNotJSONExt(fullFileName);
    const fullStackPath = path.normalize(`${stackRootPath}/${fullFileName}`);
    let stackContent = deployment.stacks[stackFileName];
    if (typeof stackContent === 'string') {
      stackContent = JSON.parse(stackContent);
    }
    await CloudformationProviderFacade.prePushCfnTemplateModifier(context, stackContent);
    fs.writeFileSync(fullStackPath, JSONUtilities.stringify(stackContent));
  }

  // Write any functions to disk
  const functionNames = Object.keys(deployment.functions);
  const functionRootPath = path.normalize(`${directory}/functions`);
  if (!fs.existsSync(functionRootPath)) {
    fs.mkdirSync(functionRootPath);
  }
  for (const functionName of functionNames) {
    const fullFunctionPath = path.normalize(`${functionRootPath}/${functionName}`);
    const zipContents = fs.readFileSync(deployment.functions[functionName]);
    fs.writeFileSync(fullFunctionPath, zipContents);
  }
  const { rootStack } = deployment;
  const rootStackPath = path.normalize(`${directory}/${rootStackFileName}`);
  const rootStackString = JSON.stringify(rootStack, null, 4);
  fs.writeFileSync(rootStackPath, rootStackString);

  // Write params to disk
  const jsonString = JSON.stringify(buildParameters, null, 4);
  const parametersOutputFilePath = path.join(directory, PARAMETERS_FILE_NAME);
  fs.writeFileSync(parametersOutputFilePath, jsonString);
}

function initStacksAndResolversDirectories(directory: string) {
  const resolverRootPath = resolverDirectoryPath(directory);
  if (!fs.existsSync(resolverRootPath)) {
    fs.mkdirSync(resolverRootPath);
  }
  const stackRootPath = stacksDirectoryPath(directory);
  if (!fs.existsSync(stackRootPath)) {
    fs.mkdirSync(stackRootPath);
  }
}

function pipelineFunctionDirectoryPath(rootPath: string) {
  return path.normalize(path.join(rootPath, 'pipelineFunctions'));
}

function resolverDirectoryPath(rootPath: string) {
  return path.normalize(`${rootPath}/resolvers`);
}

function stacksDirectoryPath(rootPath: string) {
  return path.normalize(`${rootPath}/stacks`);
}

function throwIfNotJSONExt(stackFile: string) {
  const extension = path.extname(stackFile);
  if (extension === '.yaml' || extension === '.yml') {
    throw new Error(`Yaml is not yet supported. Please convert the CloudFormation stack ${stackFile} to json.`);
  }
  if (extension !== '.json') {
    throw new Error(`Invalid extension ${extension} for stack ${stackFile}`);
  }
}

const emptyBuildDirPreserveTsconfig = (directory: string) => {
  const files = fs.readdirSync(directory);
  files.forEach((file) => {
    const fileDir = path.join(directory, file);
    if (fs.lstatSync(fileDir).isDirectory()) {
      rimraf.sync(fileDir);
    } else if (!file.endsWith('tsconfig.resource.json')) {
      fs.unlinkSync(fileDir);
    }
  });
};

export const getDeltaSyncTableTtl = (resourceOverrides: AmplifyApiGraphQlResourceStackTemplate, resource: TransformerResolverProvider): number => {
  if (_.get(resource, 'typeName') !== 'Query') {
    return SyncUtils.syncDataSourceConfig().DeltaSyncTableTTL;
  }
  const modelName = _.get(resource, ['datasource', 'name'])?.replace(new RegExp('Table$'), '');
  const deltaSyncTtlOverride = _.get(resourceOverrides, ['models', modelName, 'modelDatasource', 'dynamoDbConfig', 'deltaSyncConfig', 'deltaSyncTableTtl']);
  return deltaSyncTtlOverride || SyncUtils.syncDataSourceConfig().DeltaSyncTableTTL;
}

export const stacksTypes: Record<string, string> = {
  API: 'api',
  MODELS: 'models',
  HttpStack: 'http',
  FunctionDirectiveStack: 'function',
  PredictionsDirectiveStack: 'predictions',
  SearchableStack: 'openSearch',
};

const rootStackNameInConstruct = 'transformer-root-stack';

export const getStackMeta = (constructPathArr: string[], id: string, nestedStackArr: string[], node: Construct): ConstructResourceMeta => {
  const resource = node as CfnResource;
  if (nestedStackArr.find(val => val === constructPathArr[1])) {
    const nestedStackName = nestedStackArr.find(val => val === constructPathArr[1]);
    const resourceName = constructPathArr.filter(path => path !== nestedStackName && path !== rootStackNameInConstruct).join('');
    return {
      resourceName,
      resourceType: resource.cfnResourceType,
      nestedStack: {
        stackName: nestedStackName!,
        stackType: stacksTypes[nestedStackName!] ?? stacksTypes['MODELS'],
      },
    };
  } else {
    // root stack
    const resourceName = constructPathArr.filter(path => path !== rootStackNameInConstruct).join('');
    return {
      resourceName: id === 'Resource' ? resourceName : `${resourceName}${id}`,
      resourceType: resource.cfnResourceType,
      rootStack: {
        stackName: constructPathArr[0],
        stackType: stacksTypes.API,
      },
    };
  }
};

export const convertToAppsyncResourceObj = (amplifyObj: any) => {
  let appsyncResourceObject: AmplifyApiGraphQlResourceStackTemplate = {};
  Object.keys(amplifyObj).forEach(keys => {
    if (keys === 'api') {
      appsyncResourceObject.api = amplifyObj.api;
    } else if (keys === 'models' && !_.isEmpty(amplifyObj[keys])) {
      // require filter using keyName
      appsyncResourceObject.models = {};
      Object.keys(amplifyObj.models).forEach(key => {
        appsyncResourceObject.models![key] = generateModelDirectiveObject(amplifyObj.models[key]);
      });
    } else if (keys === 'function' && !_.isEmpty(amplifyObj[keys])) {
      const functionStackObj = amplifyObj.function.FunctionDirectiveStack;
      appsyncResourceObject.function = generateFunctionDirectiveObject(functionStackObj);
    } else if (keys === 'http' && !_.isEmpty(amplifyObj[keys])) {
      const httpStackObj = amplifyObj.http.HttpStack;
      appsyncResourceObject.http = generateHttpDirectiveObject(httpStackObj);
    } else if (keys === 'openSearch' && !_.isEmpty(amplifyObj[keys])) {
      const openSearchStackObj = amplifyObj.openSearch.SearchableStack;
      appsyncResourceObject.opensearch = generateOpenSearchDirectiveObject(openSearchStackObj);
    } else if (keys === 'predictions' && !_.isEmpty(amplifyObj[keys])) {
      appsyncResourceObject.predictions = amplifyObj.predictions.PredictionsDirectiveStack;
      if (!_.isEmpty(amplifyObj.predictions.PredictionsDirectiveStack['predictionsLambda.handler'])) {
        appsyncResourceObject.predictions!.predictionsLambdaFunction =
          amplifyObj.predictions.PredictionsDirectiveStack['predictionsLambda.handler'];
      }
    }
  });
  return appsyncResourceObject;
};

const generateFunctionDirectiveObject = (functionStackObj: any) => {
  let functionObj: Partial<FunctionDirectiveStack & AppsyncStackCommon> = {};
  Object.keys(functionStackObj).forEach(key => {
    if (key.endsWith('resolvers')) {
      functionObj.resolvers = functionStackObj.resolvers;
    } else if (key.endsWith('appsyncFunctions')) {
      functionObj.appsyncFunctions = functionStackObj.appsyncFunctions;
    } else if (functionStackObj[key].cfnResourceType.includes('DataSource')) {
      if (!functionObj.lambdaDataSource) {
        functionObj.lambdaDataSource = {};
      }
      const name = key.substring(0, key.indexOf('LambdaDataSource'));
      functionObj.lambdaDataSource[name] = functionStackObj[key];
    } else if (functionStackObj[key].cfnResourceType.includes('Role')) {
      if (!functionObj.lambdaDataSourceRole) {
        functionObj.lambdaDataSourceRole = {};
      }
      const name = key.substring(0, key.indexOf('LambdaDataSourceServiceRole'));
      functionObj.lambdaDataSourceRole[name] = functionStackObj[key];
    } else if (functionStackObj[key].cfnResourceType.includes('Policy')) {
      if (!functionObj.lambdaDataSourceServiceRoleDefaultPolicy) {
        functionObj.lambdaDataSourceServiceRoleDefaultPolicy = {};
      }
      const name = key.substring(0, key.indexOf('LambdaDataSourceServiceRoleDefaultPolicy'));
      functionObj.lambdaDataSourceServiceRoleDefaultPolicy[name] = functionStackObj[key];
    }
  });
  return functionObj;
};

const generateHttpDirectiveObject = (httpStackObj: any) => {
  let httpObj: Partial<HttpsDirectiveStack & AppsyncStackCommon> = {};
  Object.keys(httpStackObj).forEach(key => {
    if (key.endsWith('resolvers')) {
      httpObj.resolvers = httpStackObj.resolvers;
    } else if (key.endsWith('appsyncFunctions')) {
      httpObj.appsyncFunctions = httpStackObj.appsyncFunctions;
    } else if (httpStackObj[key].cfnResourceType.includes('DataSource')) {
      if (!httpObj.httpsDataSource) {
        httpObj.httpsDataSource = {};
      }
      const name = key.substring(0, key.indexOf('DataSource'));
      httpObj.httpsDataSource[name] = httpStackObj[key];
    } else if (httpStackObj[key].cfnResourceType.includes('Role')) {
      if (!httpObj.httpDataSourceServiceRole) {
        httpObj.httpDataSourceServiceRole = {};
      }
      const name = key.substring(0, key.indexOf('DataSourceServiceRole'));
      httpObj.httpDataSourceServiceRole[name] = httpStackObj[key];
    }
  });
  return httpObj;
};

const generateOpenSearchDirectiveObject = (opensearchStackObj: any) => {
  let opensearchObj: OpenSearchDirectiveStack & AppsyncStackCommon = _.pick(
    opensearchStackObj,
    'OpenSearchDataSource',
    'OpenSearchAccessIAMRole',
    'OpenSearchAccessIAMRoleDefaultPolicy',
    'OpenSearchDomain',
    'OpenSearchStreamingLambdaIAMRole',
    'OpenSearchStreamingLambdaIAMRoleDefaultPolicy',
    'CloudwatchLogsAccess',
    'OpenSearchStreamingLambdaFunction',
    'resolvers',
    'appsyncFunctions',
  );

  Object.keys(opensearchStackObj).forEach(key => {
    if (key !== 'resolvers' && key !== 'appsyncFunctions' && opensearchStackObj[key].cfnResourceType.includes('EventSourceMapping')) {
      if (!opensearchObj.OpenSearchModelLambdaMapping) {
        opensearchObj.OpenSearchModelLambdaMapping = {};
      }
      // filter ModelName fromm logicalID
      const name = key.substring(0, key.indexOf('LambdaMapping'));
      const modeName = key.substring('Searchable'.length, name.length);
      opensearchObj.OpenSearchModelLambdaMapping[modeName] = opensearchStackObj[key];
    }
  });
  return opensearchObj;
};

const generateModelDirectiveObject = (modelStackObj: any) => {
  let modelObj: ModelDirectiveStack = _.pick(modelStackObj, 'appsyncFunctions', 'DynamoDBAccess', 'InvokdeLambdaFunction', 'resolvers');
  let strippedModelObj = _.omit(modelStackObj, 'appsyncFunctions', 'DynamoDBAccess', 'InvokdeLambdaFunction', 'resolvers');
  Object.keys(strippedModelObj).forEach(key => {
    if (strippedModelObj[key].cfnResourceType.includes('DataSource')) {
      modelObj.modelDatasource = modelStackObj[key];
    }
    if (strippedModelObj[key].cfnResourceType.includes('Role')) {
      modelObj.modelIamRole = modelStackObj[key];
    }
    if (strippedModelObj[key].cfnResourceType.includes('Policy')) {
      modelObj.modelIamRoleDefaultPolicy = modelStackObj[key];
    }
    if (strippedModelObj[key].cfnResourceType.includes('Table')) {
      modelObj.modelDDBTable = modelStackObj[key];
    }
  });
  return modelObj;
};
