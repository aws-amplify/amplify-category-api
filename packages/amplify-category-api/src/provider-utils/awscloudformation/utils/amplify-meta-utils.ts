import * as path from 'path';
import { $TSMeta, AmplifyCategories, AmplifySupportedService, stateManager, pathManager } from '@aws-amplify/amplify-cli-core';
import _ from 'lodash';

export const authConfigHasApiKey = (authConfig?: any) => {
  if (!authConfig) {
    return false;
  }
  return (
    Array.of(authConfig.defaultAuthentication)
      .concat(authConfig.additionalAuthenticationProviders)
      .filter((auth) => !!auth) // filter out undefined elements which can happen if there are no addtl auth providers
      .map((auth) => auth.authenticationType)
      .findIndex((authType) => authType === 'API_KEY') > -1
  );
};

export const checkIfAuthExists = () => {
  const amplifyMeta = stateManager.getMeta();
  let authResourceName;
  const authServiceName = AmplifySupportedService.COGNITO;
  const authCategoryName = AmplifyCategories.AUTH;

  if (amplifyMeta[authCategoryName] && Object.keys(amplifyMeta[authCategoryName]).length > 0) {
    const categoryResources = amplifyMeta[authCategoryName];
    Object.keys(categoryResources).forEach((resource) => {
      if (categoryResources[resource].service === authServiceName) {
        authResourceName = resource;
      }
    });
  }

  return authResourceName;
};

export const getAppSyncAPINames = () => {
  return Object.entries(stateManager.getMeta()?.api || {})
    .filter(([_, apiResource]) => (apiResource as any).service === 'AppSync')
    .map(([name]) => name);
};

export const ensureNoAppSyncAPIExists = () => {
  const apiNames = getAppSyncAPINames();
  // This restriction of having a single AppSync API might change in future.
  if (apiNames?.length > 0) {
    throw new Error(
      `You already have an AppSync API named ${apiNames[0]} in your project. Use the "amplify update api" command to update your existing AppSync API.`,
    );
  }
};

export const getAppSyncAPIName = () => {
  const apiNames = getAppSyncAPINames();
  // This restriction of having a single AppSync API might change in future.
  if (apiNames?.length === 0) {
    throw new Error(`You do not have AppSync API added. Use "amplify add api" or "amplify import api" to add one to your project.`);
  }
  return apiNames[0];
};

export const getAPIResourceDir = (apiName: string) => {
  return path.join(pathManager.getBackendDirPath(), AmplifyCategories.API, apiName);
};

// some utility functions to extract the AppSync API name and config from amplify-meta

export const getAppSyncAuthConfig = (projectMeta: $TSMeta) => {
  const entry = getAppSyncAmplifyMetaEntry(projectMeta);
  if (entry) {
    const value = entry[1] as any;
    return value && value.output ? value.output.authConfig : {};
  }
};

export const getAppSyncResourceName = (projectMeta: $TSMeta): string | undefined => {
  const entry = getAppSyncAmplifyMetaEntry(projectMeta);
  if (entry) {
    return entry[0];
  }
};

// project meta is the contents of amplify-meta.json
// typically retreived using context.amplify.getProjectMeta()
const getAppSyncAmplifyMetaEntry = (projectMeta: $TSMeta) => {
  return Object.entries(projectMeta[AmplifyCategories.API] || {}).find(
    ([, value]) => (value as Record<string, any>).service === AmplifySupportedService.APPSYNC,
  );
};
