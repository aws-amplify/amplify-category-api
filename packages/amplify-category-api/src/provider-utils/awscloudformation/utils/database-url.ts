import { URL } from 'url';
import { ImportedRDSType, RDSDataSourceConfig } from '@aws-amplify/graphql-transformer-core';
import { AmplifyError } from '@aws-amplify/amplify-cli-core';

export const parseDatabaseUrl = (databaseUrl: string): Partial<RDSDataSourceConfig> => {
  const allowedProtocols = ['mysql', 'mysql2'];
  try {
    const parsedDatabaseUrl = new URL(databaseUrl);
    const { username, password, hostname: host } = parsedDatabaseUrl;
    const database = parsedDatabaseUrl?.pathname?.slice(1);
    const port = parseInt(parsedDatabaseUrl?.port, 10);
    const engine = parsedDatabaseUrl?.protocol?.slice(0, -1) as ImportedRDSType;

    const isValidEngine = allowedProtocols.includes(engine);
    if (!isValidEngine) {
      throw new AmplifyError('InputValidationError', {
        message: `Invalid engine ${engine}.`,
      });
    }

    const config = {
      engine,
      username,
      password,
      database,
      host,
      port,
    };
    return config;
  } catch (err) {
    if (err.code !== 'ERR_INVALID_URL') {
      throw err;
    }
  }
  return {};
};
