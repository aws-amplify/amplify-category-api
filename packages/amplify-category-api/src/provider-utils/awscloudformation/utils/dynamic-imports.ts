import * as path from 'path';

/**
 *
 * @param service
 */
export const serviceMetadataFor = async (service: string) => (await import(path.join('..', '..', 'supported-services'))).supportedServices[service];

/**
 *
 * @param datasource
 */
export const datasourceMetadataFor = async (datasource: string) => (await import(path.join('..', '..', 'supported-datasources'))).supportedDataSources[datasource];

/**
 *
 * @param walkthroughFilename
 */
export const getServiceWalkthrough = async (walkthroughFilename: string) => (await import(path.join('..', 'service-walkthroughs', walkthroughFilename))).serviceWalkthrough;
