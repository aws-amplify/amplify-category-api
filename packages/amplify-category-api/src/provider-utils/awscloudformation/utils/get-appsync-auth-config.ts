import { $TSContext } from '@aws-amplify/amplify-cli-core';
import { AppsyncApiInputState } from '../api-input-manager/appsync-api-input-state';
import { appSyncAuthTypeToAuthConfig } from './auth-config-to-app-sync-auth-type-bi-di-mapper';

/**
 *
 * @param context
 * @param resourceName
 * @returns authConfig
 */
export const getAuthConfig = async (context: $TSContext, resourceName: string) => {
  const cliState = new AppsyncApiInputState(context, resourceName);
  if (cliState.cliInputFileExists()) {
    const appsyncInputs = cliState.getCLIInputPayload().serviceConfiguration;
    return {
      defaultAuthentication: appSyncAuthTypeToAuthConfig(appsyncInputs.defaultAuthType),
      additionalAuthenticationProviders: (appsyncInputs.additionalAuthTypes || []).map(appSyncAuthTypeToAuthConfig),
    };
  }
};
