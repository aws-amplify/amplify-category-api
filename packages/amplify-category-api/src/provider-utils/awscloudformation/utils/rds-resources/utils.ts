import { parse, FieldDefinitionNode, ObjectTypeDefinitionNode, visit } from 'graphql';
import _ from 'lodash';
import { isSqlStrategy } from '@aws-amplify/graphql-transformer-core';
import {
  DataSourceStrategiesProvider,
  ModelDataSourceStrategy,
  SqlDirectiveDataSourceStrategy,
} from '@aws-amplify/graphql-transformer-interfaces';

export const checkForUnsupportedDirectives = (schema: string, context: DataSourceStrategiesProvider): void => {
  const unsupportedRDSDirectives = ['searchable', 'predictions', 'function', 'manyToMany', 'http', 'mapsTo'];
  if (_.isEmpty(schema) || _.isEmpty(context.dataSourceStrategies)) {
    return;
  }

  const rdsModels = Object.entries(context.dataSourceStrategies)
    .filter(([key, value]) => isSqlStrategy(value))
    .map(([key, value]) => key);

  if (_.isEmpty(rdsModels)) {
    return;
  }

  const document = parse(schema);
  const schemaVisitor = {
    FieldDefinition: {
      enter(node: FieldDefinitionNode, key, parent, path, ancestors): any {
        const parentName = getParentName(ancestors);
        if (!(parentName === 'Query') && !rdsModels?.includes(parentName)) {
          return;
        }
        node?.directives?.map((directive) => {
          if (unsupportedRDSDirectives.includes(directive?.name?.value)) {
            throw unsupportedDirectiveError(directive?.name?.value, node?.name?.value, parentName, unsupportedRDSDirectives);
          }
        });
      },
    },
    ObjectTypeDefinition: {
      enter(node: ObjectTypeDefinitionNode): any {
        const typeName = node?.name?.value;
        if (!(typeName === 'Query') && !rdsModels?.includes(typeName)) {
          return;
        }
        node?.directives?.map((directive) => {
          if (unsupportedRDSDirectives.includes(directive?.name?.value)) {
            throw unsupportedDirectiveError(directive?.name?.value, undefined, node?.name?.value, unsupportedRDSDirectives);
          }
        });
      },
    },
  };

  visit(document, schemaVisitor);
};

export const containsSqlModelOrDirective = (
  dataSourceStrategies: Record<string, ModelDataSourceStrategy>,
  sqlDirectiveDataSourceStrategies?: SqlDirectiveDataSourceStrategy[],
): boolean => {
  if (sqlDirectiveDataSourceStrategies && sqlDirectiveDataSourceStrategies?.length > 0) {
    return true;
  }
  return Object.values(dataSourceStrategies).some((strategy) => isSqlStrategy(strategy));
};

const unsupportedDirectiveError = (
  directiveName: string,
  fieldName: string | undefined,
  typeName: string,
  unsupportedDirectives: string[],
): Error => {
  return new Error(
    `@${directiveName} directive on type "${typeName}" ${
      fieldName ? `and field "${fieldName}"` : ''
    } is not supported on a SQL datasource. Following directives are not supported on a SQL datasource: ${unsupportedDirectives.join(
      ', ',
    )}`,
  );
};

const getParentName = (ancestors: any[]): string | undefined => {
  if (ancestors && ancestors?.length > 0) {
    return (ancestors[ancestors.length - 1] as ObjectTypeDefinitionNode)?.name?.value;
  }
  return undefined;
};
