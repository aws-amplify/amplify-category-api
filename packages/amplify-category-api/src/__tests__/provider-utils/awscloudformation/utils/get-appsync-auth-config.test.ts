import { $TSContext } from '@aws-amplify/amplify-cli-core';
import { getAuthConfig } from '../../../../provider-utils/awscloudformation/utils/get-appsync-auth-config';

const getCLIInputPayload_mock = jest
  .fn()
  .mockReturnValueOnce({
    version: 1,
    serviceConfiguration: {
      apiName: 'authv2migration1',
      serviceName: 'AppSync',
      defaultAuthType: {
        mode: 'AWS_IAM',
      },
      conflictResolution: {},
      additionalAuthTypes: [],
    },
  })
  .mockReturnValueOnce({
    version: 1,
    serviceConfiguration: {
      apiName: 'authv2migration1',
      serviceName: 'AppSync',
      defaultAuthType: {
        mode: 'AWS_IAM',
      },
      conflictResolution: {},
      additionalAuthTypes: [
        {
          mode: 'API_KEY',
          expirationTime: 7,
          keyDescription: '',
        },
      ],
    },
  });

jest.mock('../../../../provider-utils/awscloudformation/api-input-manager/appsync-api-input-state.ts', () => {
  return {
    AppsyncApiInputState: jest.fn().mockImplementation(() => {
      return {
        getCLIInputPayload: getCLIInputPayload_mock,
        cliInputFileExists: jest.fn().mockReturnValue(true),
      };
    }),
  };
});

const mockContext: $TSContext = {
  amplify: {
    getCategoryPluginInfo: (_context: $TSContext, category: string) => {
      return {
        packageLocation: `@aws-amplify/amplify-category-${category}`,
      };
    },
  },
  input: {
    options: {},
  },
} as unknown as $TSContext;

test('function with default auth config', async () => {
  expect(await getAuthConfig(mockContext, 'mockapiResource')).toMatchSnapshot();
});

test('function with default and additional auth config', async () => {
  expect(await getAuthConfig(mockContext, 'mockapiResource')).toMatchSnapshot();
});
