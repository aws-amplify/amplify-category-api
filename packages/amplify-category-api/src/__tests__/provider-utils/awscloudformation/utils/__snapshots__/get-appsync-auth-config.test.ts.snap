// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function with default and additional auth config 1`] = `
Object {
  "additionalAuthenticationProviders": Array [
    Object {
      "apiKeyConfig": Object {
        "apiKeyExpirationDate": undefined,
        "apiKeyExpirationDays": 7,
        "description": "",
      },
      "authenticationType": "API_KEY",
    },
  ],
  "defaultAuthentication": Object {
    "authenticationType": "AWS_IAM",
  },
}
`;

exports[`function with default auth config 1`] = `
Object {
  "additionalAuthenticationProviders": Array [],
  "defaultAuthentication": Object {
    "authenticationType": "AWS_IAM",
  },
}
`;
